application/models/services/QuestionGroupService.php

```diff
@@ -0,0 +1,410 @@
+     *                     ['questionGroup']
+     *                          [gid]
+     *                          [sid]
+     *                          [group_order]
+     *                          [randomization_group]
+     *                          [grelevance]
+     *                     ['questionGroupI10N']
+     *                         [en]
+     *                            [group_name]
+     *                            [description]
+     *                         [...]    //more languages
+     * @return QuestionGroup
+     */
+    public function updateGroup(int $surveyId, int $questionGroupId, array $input)
+    {
+        $survey = $this->getSurvey($surveyId);
+
+        if (!$this->modelPermission->hasSurveyPermission($survey->sid, 'surveycontent', 'update')) {
+            throw new Exception(
+                'Permission denied'
+            );
+        }
+
+        $questionGroup = $this->modelQuestionGroup->findByPk($questionGroupId);
+        if (!$questionGroup) {
+            throw new Exception(
+                'Group not found'
+            );
+        }
+
+        $questionGroup = $this->updateQuestionGroup($questionGroup, $input['questionGroup']);
+        $this->updateQuestionGroupLanguages($questionGroup, $input['questionGroupI10N']);
+
+        return $questionGroup;
+    }
+
+    /**
+     * Creates a question group and all the languages.
+     *
+     * @param int $surveyId the survey id
+     * @param array $input has the data for a question group, including an array for languages
+     *                     ['questionGroup']
+     *                          [gid]
+     *                          [sid]
+     *                          [group_order]
+     *                          [randomization_group]
+     *                          [grelevance]
+     *                     ['questionGroupI10N']
+     *                         [en]
+     *                            [group_name]
+     *                            [description]
+     *                         [...]    //more languages
+     * @return QuestionGroup
+     */
+    public function createGroup(int $surveyId, array $input)
+    {
+        if (!$this->modelPermission->hasSurveyPermission($surveyId, 'surveycontent', 'update')) {
+            throw new Exception(
+                'Permission denied'
+            );
+        }
+
+        $questionGroup = $this->newQuestionGroup($surveyId, $input['questionGroup']);
+        $this->updateQuestionGroupLanguages($questionGroup, $input['questionGroupI10N']);
+
+        return $questionGroup;
+    }
+
+    /**
+     * Deletes a question group and all its dependencies.
+     *
+     * @param int $questionGroupId the question group id
+     * @param int $surveyId the survey id
+     * @return int|null number of deleted rows
+     * @throws CHttpException
+     */
+    public function deleteGroup(int $questionGroupId, int $surveyId)
```

Comment:
If you pass in QuestionGroup model directly instead of just the id, you might not need to inject the QuestionGroup modal in the constructor. Again, @kevin-foster-uk will probably not agree. Just saying, having so many dependencies in the constructor when they can be given as method arguments instead might be preferable at some point.

Reply:
I agree that are ways the code can be improved. That isn't the focus on this round of refactoring. 

As I have said before. This refactoring round is not intended to fix all code smell. We have two goals:

1. Re-usability
2. Test ability

The process for refactoring has been this:

- Copy existing code into a service class
- If the class is too big break it into multiple composed classes
- Replace Yii MVC specific code with model specific code where possible (e.g. Request data, Exceptions)
- Inject dependancies to allow for testing
- Try to pass code-sniffer and psalm rules

I have not spent much time looking at the implementation of the business logic or how it could be improved other than when it was necessary in order to meet goal 1. and 2. above. The work has mostly been copying and pasting and re-arranging.

Again, I agree that are ways the code can be improved. That isn't the focus on this round of refactoring.